pkg/controllers/webhookcerts/cert_requester.go

```diff
@@ -0,0 +1,380 @@
+				}
+				// do a live Get, the cache might be stale at this point if we did an Update removing the cert before
+				certSecret, err := c.secretLister.Secrets("openshift-oauth-apiserver").Get(certSecretName)
+				if err != nil {
+					return nil, err
+				}
+				certSecret.Data["tls.crt"] = csrCert
+				_, err = c.secretClient.Update(ctx, certSecret, metav1.UpdateOptions{})
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+
+	}
+
+	if err = c.resubmitCSR(ctx, key); err != nil {
+		return nil, err
+	}
+
+	return pbool(true), nil
+}
+
+func (c *webhookAuthenticatorCertRequester) getKeyCertFromSecret(ctx context.Context) (key *rsa.PrivateKey, cert *x509.Certificate, keyRotated bool, err error) {
+	certSecret, err := c.secretLister.Secrets("openshift-oauth-apiserver").Get(certSecretName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, nil, false, err
+	}
+
+	var keyPEM []byte
+	if certSecret != nil {
+		keyPEM = certSecret.Data["tls.key"]
+	}
+	if len(keyPEM) > 0 {
+		key, err = pemDecodeRSAPrivate(keyPEM)
+		if err != nil {
+			klog.Error("failed to decode the private key used for the openshift-authenticator, new key will be generated: %v", err)
+			// delete the secret and regenerate it in the next step with a new key
+			err = c.secretClient.Delete(ctx, certSecretName, metav1.DeleteOptions{})
+			if err != nil {
+				return nil, nil, false, err
+			}
+		}
+	}
+	if key == nil {
+		key, err = rsa.GenerateKey(rand.Reader, 4096)
+		if err != nil {
+			return nil, nil, false, fmt.Errorf("failed to generate a private key: %w", err)
+		}
+		keyPEM, err := pemEncodeRSAPrivate(key)
+		if err != nil {
+			return nil, nil, false, err
+		}
+
+		if err := c.secretClient.Delete(ctx, certSecretName, metav1.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+			return nil, nil, false, err
+		}
+
+		_, err = c.secretClient.Create(ctx, &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: certSecretName,
+			},
+			Data: map[string][]byte{
+				"tls.key": keyPEM,
+			},
+		}, metav1.CreateOptions{})
+
+		return key, nil, true, err
+	}
+
+	certPEM := certSecret.Data["tls.crt"]
+	if len(certPEM) > 0 {
+		cert, err = pemDecodeCert(certPEM)
+
+		// delete tls.crt if it's malformed or its pub key does not match the private key
+		var deleteCert bool
+		if err != nil {
+			klog.Error("failed to decode the certificate used for the openshift-authenticator, new certificate will be requested: %v", err)
+			deleteCert = true
+		} else if !key.PublicKey.Equal(cert.PublicKey.(*rsa.PublicKey)) {
+			klog.Error("the public key of the certificate used for the openshift-authenticator does not match the stored private key, new certificate will be requested: %v", err)
+			deleteCert = true
+		}
+
+		if deleteCert {
+			certSecretCopy := certSecret.DeepCopy()
+			delete(certSecretCopy.Data, "tls.crt")
+			_, err := c.secretClient.Update(ctx, certSecretCopy, metav1.UpdateOptions{})
+			return key, nil, false, err
```

Comment:
should we resubmit if we deleted the cert?

Reply:
If the certificate already exists in our CSR, it's going to replace this one, otherwise we either wait for the current CSR to get a cert, or if there is no CSR, we submit a new one.